### PR TITLE
[5.1] Minor cleanup for component JS

### PR DESCRIPTION
**Pull request by C-Lodder**
### Summary of Changes

As Joomla no longer supports IE11, this PR mainly removes the need for converting a NodeList to an Array, as evergreen browser's implementation of `forEach` can iterater over a NodeList.

Also includes some other minor tweaks.

Happy to work on the remaining JS

image

### Testing Instructions

Code review @dgrammatiko @Fedik

### CODE REVIEW

File: build/media_source/com_finder/js/finder.es6.js [hunk 1]
```diff
@@ -85,9 +83,7 @@
       }
     });
 
-    const forms = [].slice.call(document.querySelectorAll('.js-finder-searchform'));
-
-    forms.forEach((form) => {
+    document.querySelectorAll('.js-finder-searchform').forEach((form) => {
       form.addEventListener('submit', onSubmit);
     });
```
Comment (dgrammatiko) [inline build/media_source/com_finder/js/finder.es6.js]:
```suggestion
    document.querySelectorAll('.js-finder-searchform').forEach((form) => form.addEventListener('submit', onSubmit));
```

File: build/media_source/com_joomlaupdate/js/default.es6.js [hunk 2]
```diff
@@ -233,22 +233,22 @@ Joomla = window.Joomla || {};
...
         const compatibilityTable = el.closest('.compatibilityTable');
 
         if (el.dataset.state === 'closed') {
           el.dataset.state = 'open';
           el.innerHTML = Joomla.sanitizeHtml(Joomla.Text._('COM_JOOMLAUPDATE_VIEW_DEFAULT_EXTENSIONS_SHOW_LESS_COMPATIBILITY_INFORMATION'));
 
-          [].slice.call(compatibilityTable.querySelectorAll('table .hidden')).forEach((elem) => {
+          compatibilityTable.querySelectorAll('table .hidden').forEach((elem) => {
             elem.classList.remove('hidden');
           });
```
Comment (dgrammatiko) [inline build/media_source/com_joomlaupdate/js/default.es6.js]:
```suggestion
          compatibilityTable.querySelectorAll('table .hidden').forEach((elem) => elem.classList.remove('hidden'));
```

File: build/media_source/com_joomlaupdate/js/default.es6.js [hunk 3]
```diff
@@ -233,22 +233,22 @@ Joomla = window.Joomla || {};
...
         } else {
           el.dataset.state = 'closed';
           el.innerHTML = Joomla.sanitizeHtml(Joomla.Text._('COM_JOOMLAUPDATE_VIEW_DEFAULT_EXTENSIONS_SHOW_MORE_COMPATIBILITY_INFORMATION'));
 
-          [].slice.call(compatibilityTable.querySelectorAll('table .instver, table .upcomp, table .currcomp')).forEach((elem) => {
+          compatibilityTable.querySelectorAll('table .instver, table .upcomp, table .currcomp').forEach((elem) => {
             elem.classList.add('hidden');
           });
```
Comment (dgrammatiko) [inline build/media_source/com_joomlaupdate/js/default.es6.js]:
```suggestion
          compatibilityTable.querySelectorAll('table .instver, table .upcomp, table .currcomp').forEach((elem) => elem.classList.add('hidden'));
```

File: build/media_source/com_joomlaupdate/js/default.es6.js [hunk 4]
```diff
@@ -258,14 +258,13 @@ Joomla = window.Joomla || {};
...
-        };
-        extensionsInformation.push(thisInfo);
-      });
+    extensions.forEach((extension) => {
+      const thisInfo = {
+        eid: extension.getAttribute('data-extension-id'),
+        version: extension.getAttribute('data-extension-current-version'),
+      };
+      extensionsInformation.push(thisInfo);
+    });
```
Comment (dgrammatiko) [inline build/media_source/com_joomlaupdate/js/default.es6.js]:
```suggestion
    extensions.forEach((extension) => extensionsInformation.push({
        eid: extension.getAttribute('data-extension-id'),
        version: extension.getAttribute('data-extension-current-version'),
      }));
```

File: build/media_source/com_languages/js/overrider.es6.js [hunk 5]
```diff
@@ -79,9 +79,7 @@
         this.states.searchType = formSearchType.value || 'value';
 
         // Remove the old results
-        const oldResults = [].slice.call(document.querySelectorAll('.language-results'));
-
-        oldResults.forEach((result) => {
+        document.querySelectorAll('.language-results').forEach((result) => {
           result.parentNode.removeChild(result);
         });
```
Comment (dgrammatiko) [inline build/media_source/com_languages/js/overrider.es6.js]:
```suggestion
        document.querySelectorAll('.language-results').forEach((result) => result.parentNode.removeChild(result));
```

File: build/media_source/com_languages/js/overrider.es6.js [hunk 6]
```diff
@@ -99,8 +97,7 @@
         // Otherwise it is a new searchstring and we have to remove all previous results first
         this.moreResults.classList.remove('show');
 
-        const childs = [].slice.call(document.querySelectorAll('#results-container div.language-results'));
-        childs.forEach((child) => {
+        document.querySelectorAll('#results-container div.language-results').forEach((child) => {
           child.parentNode.removeChild(child);
         });
```
Comment (dgrammatiko) [inline build/media_source/com_languages/js/overrider.es6.js]:
```suggestion
        document.querySelectorAll('#results-container div.language-results').forEach((child) => child.parentNode.removeChild(child));
```

File: build/media_source/com_menus/js/admin-item-edit.es6.js [hunk 7]
```diff
@@ -10,17 +10,13 @@ if (!window.Joomla) {
 Joomla.submitbutton = (task, type) => {
   if (task === 'item.setType' || task === 'item.setMenuType') {
     if (task === 'item.setType') {
-      const list = [].slice.call(document.querySelectorAll('#item-form input[name="jform[type]"]'));
-
-      list.forEach((item) => {
+      document.querySelectorAll('#item-form input[name="jform[type]"]').forEach((item) => {
         item.value = type;
       });
```
Comment (dgrammatiko) [inline build/media_source/com_menus/js/admin-item-edit.es6.js:15]:
```suggestion
      document.querySelectorAll('#item-form input[name="jform[type]"]').forEach((item) => item.value = type);
```

File: build/media_source/com_menus/js/admin-item-edit.es6.js [hunk 8]
```diff
@@ -10,17 +10,13 @@ if (!window.Joomla) {
...
 
       document.getElementById('fieldtype').value = 'type';
     } else {
-      const list = [].slice.call(document.querySelectorAll('#item-form input[name="jform[menutype]"]'));
-
-      list.forEach((item) => {
+      document.querySelectorAll('#item-form input[name="jform[menutype]"]').forEach((item) => {
         item.value = type;
       });
```
Comment (dgrammatiko) [inline build/media_source/com_menus/js/admin-item-edit.es6.js:21]:
```suggestion
      document.querySelectorAll('#item-form input[name="jform[menutype]"]').forEach((item) => item.value = type);
```

File: build/media_source/com_menus/js/admin-item-edit_container.es6.js [hunk 9]
```diff
@@ -39,7 +39,7 @@
     }
   };
 
-  [].slice.call(document.querySelectorAll('.treeselect input[type="checkbox"]')).forEach((checkbox) => {
+  document.querySelectorAll('.treeselect input[type="checkbox"]').forEach((checkbox) => {
     checkbox.addEventListener('click', switchState);
   });
```
Comment (dgrammatiko) [inline build/media_source/com_menus/js/admin-item-edit_container.es6.js]:
```suggestion
  document.querySelectorAll('.treeselect input[type="checkbox"]').forEach((checkbox) => checkbox.addEventListener('click', switchState));
```

File: build/media_source/com_templates/js/admin-template-toggle-assignment.es6.js [hunk 10]
```diff
@@ -7,8 +7,7 @@ Joomla = window.Joomla || {};
 
 ((Joomla) => {
   Joomla.toggleAll = () => {
-    const checkBoxes = [].slice.call(document.querySelectorAll('.chk-menulink'));
-    checkBoxes.forEach((checkBox) => {
+    document.querySelectorAll('.chk-menulink').forEach((checkBox) => {
       checkBox.checked = !checkBox.checked;
     });
```
Comment (dgrammatiko) [inline build/media_source/com_templates/js/admin-template-toggle-assignment.es6.js:12]:
```suggestion
    document.querySelectorAll('.chk-menulink').forEach((checkBox) => checkBox.checked = !checkBox.checked);
```

File: build/media_source/com_templates/js/admin-templates-default.es6.js [hunk 11]
```diff
@@ -52,11 +52,11 @@
         event.currentTarget.classList.add('selected');
         const ismedia = event.currentTarget.dataset.base === 'media' ? 1 : 0;
 
-        [...document.querySelectorAll('#fileModal input.address')].forEach((element) => {
+        document.querySelectorAll('#fileModal input.address').forEach((element) => {
           element.value = event.currentTarget.getAttribute('data-id');
         });
```
Comment (dgrammatiko) [inline build/media_source/com_templates/js/admin-templates-default.es6.js:57]:
```suggestion
        document.querySelectorAll('#fileModal input.address').forEach((element) => element.value = event.currentTarget.getAttribute('data-id'));
```

File: build/media_source/com_templates/js/admin-templates-default.es6.js [hunk 12]
```diff
@@ -52,11 +52,11 @@
...
 
-        [...document.querySelectorAll('#fileModal input[name="isMedia"]')].forEach((el) => {
+        document.querySelectorAll('#fileModal input[name="isMedia"]').forEach((el) => {
           el.value = ismedia;
         });
```
Comment (dgrammatiko) [inline build/media_source/com_templates/js/admin-templates-default.es6.js:61]:
```suggestion
        document.querySelectorAll('#fileModal input[name="isMedia"]').forEach((el) => el.value = ismedia);
```

File: build/media_source/com_templates/js/admin-templates-default.es6.js [hunk 13]
```diff
@@ -74,18 +74,18 @@
         event.currentTarget.classList.add('selected');
         const ismedia = event.currentTarget.dataset.base === 'media' ? 1 : 0;
 
-        [...document.querySelectorAll('#folderModal input.address')].forEach((element) => {
+        document.querySelectorAll('#folderModal input.address').forEach((element) => {
           element.value = event.currentTarget.getAttribute('data-id');
         });
```
Comment (dgrammatiko) [inline build/media_source/com_templates/js/admin-templates-default.es6.js:79]:
```suggestion
        document.querySelectorAll('#folderModal input.address').forEach((element) => element.value = event.currentTarget.getAttribute('data-id'));
```

File: build/media_source/com_templates/js/admin-templates-default.es6.js [hunk 14]
```diff
@@ -74,18 +74,18 @@
...
 
-        [...document.querySelectorAll('#folderModal input[name="isMedia"]')].forEach((el) => {
+        document.querySelectorAll('#folderModal input[name="isMedia"]').forEach((el) => {
           el.value = ismedia;
         });
```
Comment (dgrammatiko) [inline build/media_source/com_templates/js/admin-templates-default.es6.js:83]:
```suggestion
        document.querySelectorAll('#folderModal input[name="isMedia"]').forEach((el) => el.value = ismedia);
```

File: build/media_source/com_actionlogs/js/admin-actionlogs-default.es6.js [hunk 15]
```diff
@@ -14,7 +14,7 @@ Joomla.submitbutton = (task) => {
   if (task === 'actionlogs.exportSelectedLogs') {
     // Get id of selected action logs item and pass it to export form hidden input
     const cids = [];
-    const elements = [].slice.call(document.querySelectorAll("input[name='cid[]']:checked"));
+    const elements = document.querySelectorAll("input[name='cid[]']:checked");
 
     if (elements.length) {
```
Comment (dgrammatiko) [inline build/media_source/com_actionlogs/js/admin-actionlogs-default.es6.js]: This check is useless, if the array is empty the forEach won't run

File: build/media_source/com_cache/js/admin-cache-default.es6.js [hunk 16]
```diff
@@ -9,7 +9,7 @@ Joomla = window.Joomla || {};
   'use strict';
 
   document.addEventListener('DOMContentLoaded', () => {
-    [].slice.call(document.querySelectorAll('.cache-entry')).forEach((el) => {
+    document.querySelectorAll('.cache-entry').forEach((el) => {
       el.addEventListener('click', ({ currentTarget }) => {
         Joomla.isChecked(currentTarget.checked);
       });
```
Comment (dgrammatiko) [inline build/media_source/com_cache/js/admin-cache-default.es6.js]: The curly braces are reduntant

File: build/media_source/com_categories/js/shared-categories-accordion.es6.js [hunk 17]
```diff
@@ -43,6 +43,6 @@ const handleCategoryToggleButtonClick = ({ currentTarget }) => {
   target.toggleAttribute('hidden');
 };
 
-Array.from(document.querySelectorAll(buttonsSelector)).forEach((button) => {
+document.querySelectorAll(buttonsSelector).forEach((button) => {
   button.addEventListener('click', handleCategoryToggleButtonClick);
 });
```
Comment (dgrammatiko) [inline build/media_source/com_categories/js/shared-categories-accordion.es6.js]:
```suggestion
document.querySelectorAll(buttonsSelector).forEach((button) => button.addEventListener('click', handleCategoryToggleButtonClick));
```

File: build/media_source/com_content/js/articles-status.es6.js [hunk 18]
```diff
@@ -7,9 +7,7 @@
   'use strict';
 
   document.addEventListener('DOMContentLoaded', () => {
-    const elements = [].slice.call(document.querySelectorAll('.article-status'));
-
-    elements.forEach((element) => {
+    document.querySelectorAll('.article-status').forEach((element) => {
```
Comment (dgrammatiko) [inline build/media_source/com_content/js/articles-status.es6.js:10]: The curly braces are redundant

File: build/media_source/com_content/js/form-edit.es6.js [hunk 19]
```diff
@@ -22,8 +22,7 @@
 
   // Register events
   document.addEventListener('DOMContentLoaded', () => {
-    const buttons = [].slice.call(document.querySelectorAll(`[${buttonDataSelector}]`));
-    buttons.forEach((button) => {
+    document.querySelectorAll(`[${buttonDataSelector}]`).forEach((button) => {
```
Comment (dgrammatiko) [inline build/media_source/com_content/js/form-edit.es6.js:25]: Curly braces redundant

File: build/media_source/com_contenthistory/js/admin-compare-compare.es6.js [hunk 20]
```diff
@@ -40,8 +40,7 @@
   };
 
   const onBoot = () => {
-    const diffs = [].slice.call(document.querySelectorAll('.original'));
-    diffs.forEach((fragment) => {
+    document.querySelectorAll('.original').forEach((fragment) => {
```
Comment (dgrammatiko) [inline build/media_source/com_contenthistory/js/admin-compare-compare.es6.js]: Curly braces are redundant

File: build/media_source/com_cpanel/js/admin-cpanel-default.es6.js [hunk 21]
```diff
@@ -45,8 +45,7 @@ const debounce = (callback, time = 250, interval) => (...args) => clearTimeout(i
   const onBoot = () => {
     const cpanelModules = document.getElementById('content');
     if (cpanelModules) {
-      const links = [].slice.call(cpanelModules.querySelectorAll('.unpublish-module'));
-      links.forEach((link) => {
+      cpanelModules.querySelectorAll('.unpublish-module').forEach((link) => {
```
Comment (dgrammatiko) [inline build/media_source/com_cpanel/js/admin-cpanel-default.es6.js:48]: Curly braces are redundant

File: build/media_source/com_finder/js/finder-edit.es6.js [hunk 22]
```diff
@@ -10,30 +10,26 @@ Joomla = window.Joomla || {};
 
   document.addEventListener('DOMContentLoaded', () => {
     // Handle toggle all
-    [].slice.call(document.querySelectorAll('.filter-toggle-all')).forEach((button) => {
+    document.querySelectorAll('.filter-toggle-all').forEach((button) => {
       button.addEventListener('click', () => {
-        [].slice.call(document.querySelectorAll('.filter-node')).forEach((node) => {
-          node.click();
-        });
+        document.querySelectorAll('.filter-node').forEach((node) => node.click());
```
Comment (dgrammatiko) [inline build/media_source/com_finder/js/finder-edit.es6.js]: Curly braces are redundant

**Review by dgrammatiko — COMMENTED** [review]
Left some comments about the redundant curly braces.

Great stuff Charlie.

Are you up to also remove the redundant `DOMContentLoaded` events on the files (almost all) with either `defer` attribute or `type=module`?

**Comment by C-Lodder** [comment]
@dgrammatiko 

> Are you up to also remove the redundant DOMContentLoaded events on the files (almost all) with either defer attribute or type=module?

I'll do that in a separate PR, as it will probably involve more than just a code review.

File: build/media_source/com_cache/js/admin-cache-default.es6.js [hunk 23]
```diff
@@ -9,10 +9,8 @@ Joomla = window.Joomla || {};
...
 
   document.addEventListener('DOMContentLoaded', () => {
-    [].slice.call(document.querySelectorAll('.cache-entry')).forEach((el) => {
-      el.addEventListener('click', ({ currentTarget }) => {
-        Joomla.isChecked(currentTarget.checked);
-      });
-    });
+    document.querySelectorAll('.cache-entry').forEach((el) =>
+      el.addEventListener('click', ({ currentTarget }) => Joomla.isChecked(currentTarget.checked));
+    );
```
Comment (dgrammatiko) [inline build/media_source/com_cache/js/admin-cache-default.es6.js]:
```suggestion
    document.querySelectorAll('.cache-entry').forEach((el) =>
      el.addEventListener('click', ({ currentTarget }) => Joomla.isChecked(currentTarget.checked)));
```

File: build/media_source/com_cpanel/js/admin-cpanel-default.es6.js [hunk 24]
```diff
@@ -45,10 +45,9 @@ const debounce = (callback, time = 250, interval) => (...args) => clearTimeout(i
   const onBoot = () => {
     const cpanelModules = document.getElementById('content');
     if (cpanelModules) {
-      const links = [].slice.call(cpanelModules.querySelectorAll('.unpublish-module'));
-      links.forEach((link) => {
+      cpanelModules.querySelectorAll('.unpublish-module').forEach((link) =>
         link.addEventListener('click', ({ target }) => Joomla.unpublishModule(target));
```
Comment (dgrammatiko) [inline build/media_source/com_cpanel/js/admin-cpanel-default.es6.js:49]:
```suggestion
        link.addEventListener('click', ({ target }) => Joomla.unpublishModule(target))
```

Comment (C-Lodder) [reply build/media_source/com_cpanel/js/admin-cpanel-default.es6.js:49]: I think the linter will complain about the missing semicolon

Comment (dgrammatiko) [reply build/media_source/com_cpanel/js/admin-cpanel-default.es6.js:49]: It complains right now because it's wrong syntax Now: `););` after this: `));`

Comment (dgrammatiko) [reply build/media_source/com_cpanel/js/admin-cpanel-default.es6.js:49]: Probably the return after this parenthesis needs to also be removed

**Comment by dgrammatiko** [comment]
@C-Lodder can you fix the linter parts? Then I'll give it a test
Screenshot 2024-02-11 at 12 08 46

**Comment by dgrammatiko** [comment]
I have tested this item :white_check_mark: successfully on 29f887dcbb4eeb3ffbfd73b24e77322cc3f9020b<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42778">issues.joomla.org/tracker/joomla-cms/42778</a>.</sub>

**Comment by dgrammatiko** [comment]
@Fedik could you review this and give it a successful test?

**Comment by Fedik** [comment]
I have tested this item :white_check_mark: successfully on 29f887dcbb4eeb3ffbfd73b24e77322cc3f9020b<br /><br />Looks good on review.
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42778">issues.joomla.org/tracker/joomla-cms/42778</a>.</sub>

**Comment by Fedik** [comment]
rtc<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42778">issues.joomla.org/tracker/joomla-cms/42778</a>.</sub>

**Comment by LadySolveig** [comment]
Thank you @C-Lodder  and also for review and testing @dgrammatiko @Fedik

**Pull request merged**